ArduCopter/Analyse_Noise.cpp

```diff
@@ -0,0 +1,312 @@
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLE", 1, Analyse_Noise, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: FREQ
+    // @DisplayName: Sampling Frequency
+    // @Description: Sampling Frequency in Hz
+    // @Range: 10 2000
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("SAMPLE_FREQ", 2, Analyse_Noise, _fft_sampling_rate_hz, 1000),
+
+    // @Param: MINHZ
+    // @DisplayName: Minimum Frequency
+    // @Description: Lower bound of notch center frequency in Hz
+    // @Range: 10 400
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("MINHZ", 3, Analyse_Noise, _dyn_notch_min_hz, 80),
+
+    AP_GROUPEND
+};
+
+// The FFT splits the frequency domain into an number of bins
+// A sampling frequency of 1000 and max frequency of 500 at a window size of 32 gives 16 frequency bins each 31.25Hz wide
+// Eg [0,31), [31,62), [62, 93) etc
+// for gyro loop >= 4KHz, sample rate 2000 defines FFT range to 1000Hz, 16 bins each 62.5 Hz wide
+// NB  FFT_WINDOW_SIZE is set to 32 in gyroanalyse.h
+#define FFT_BIN_COUNT (FFT_WINDOW_SIZE / 2)
+// smoothing frequency for FFT centre frequency
+#define DYN_NOTCH_SMOOTH_FREQ_HZ 50
+// we need 4 steps for each axis
+#define DYN_NOTCH_CALC_TICKS (XYZ_AXIS_COUNT * 4)
+
+Analyse_Noise::Analyse_Noise()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void Analyse_Noise::init(uint32_t target_looptime_us, AP_InertialSensor& ins)
+{
+    _ins = &ins;
+    // If we get at least 3 samples then use the default FFT sample frequency
+    // otherwise we need to calculate a FFT sample frequency to ensure we get 3 samples (gyro loops < 4K)
+    const int gyro_loop_rate_hz = lrintf((1.0f / target_looptime_us) * 1e6f);
+
+    _fft_sampling_rate_hz = MIN(gyro_loop_rate_hz, _fft_sampling_rate_hz);
+    _fft_resolution = (float)_fft_sampling_rate_hz / FFT_WINDOW_SIZE;
+    _fft_start_bin = _dyn_notch_min_hz / lrintf(_fft_resolution);
+    _dyn_notch_max_ctr_hz = _fft_sampling_rate_hz / 2; //Nyquist
+
+    for (int i = 0; i < FFT_WINDOW_SIZE; i++) {
+        _hanning_window[i] = (0.5f - 0.5f * cosf(2 * M_PI * i / (FFT_WINDOW_SIZE - 1)));
+    }
+
+    analyse_init(target_looptime_us);
+}
+
+// a function called by the main thread at the main loop rate:
+void Analyse_Noise::periodic()
+{
+    if (!_enable) {
+        return;
+    }
+    push_sample(_ins->get_filtered_gyro());
+}
+
+void Analyse_Noise::analyse_init(uint32_t target_looptime_us)
+{
+#if CONFIG_HAL_BOARD != HAL_BOARD_SITL
+    arm_rfft_fast_init_f32(&_fft_instance, FFT_WINDOW_SIZE);
+#endif
+    for (uint8_t axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+        // any init value
+        _center_freq_hz[axis] = _dyn_notch_max_ctr_hz;
+        _prev_center_freq_hz[axis] = _dyn_notch_max_ctr_hz;
+        _detected_frequency_filter[axis].set_cutoff_frequency(_fft_sampling_rate_hz, DYN_NOTCH_SMOOTH_FREQ_HZ);
+    }
+}
+
+void Analyse_Noise::push_sample(const Vector3f& sample)
+{
```

Comment:
@andyp1per this is the function which is called by periodic (called at 400 hz from the main loop).  these are single samples right?  so you are only able to collect data at a maximum of 400 hz?

Reply:
Well the loop rate which is 800Hz in my case. If I can get it producing results that look sane I might think about integrating with the gyro loop - but even then that's basically a max rate of 1000Hz (even with fast sampling on - we average samples to the backend rate), so of dubious benefit.